pkg/agent/api/grpc/hyperstart.proto

```diff
@@ -0,0 +1,135 @@
+syntax = "proto3";
+
+package grpc;
+
+import "oci.proto";
+import "google/protobuf/empty.proto";
+
+// To generate hyperstart.pb.go, run:
+// $ go get -u github.com/google/protobuf
+// $ protoc -I=$GOPATH/src/github.com/google/protobuf/src/ -I. --go_out=plugins=grpc:. hyperstart.proto oci.proto
+
+// unstable
+service HyperstartService {
+	// execution
+	rpc CreateContainer(CreateContainerRequest) returns (google.protobuf.Empty);
+	rpc StartContainer(StartContainerRequest) returns (google.protobuf.Empty);
+	rpc ExecProcess(ExecProcessRequest) returns (google.protobuf.Empty);
+	rpc SignalProcess(SignalProcessRequest) returns (google.protobuf.Empty);
+	rpc WaitProcess(WaitProcessRequest) returns (WaitProcessResponse); // wait & reap like waitpid(2)
+
+	// stdio
+	rpc WriteStdin(WriteStreamRequest) returns (WriteStreamResponse);
+	rpc ReadStdout(ReadStreamRequest) returns (ReadStreamResponse);
+	rpc ReadStderr(ReadStreamRequest) returns (ReadStreamResponse);
+	rpc CloseStdin(CloseStdinRequest) returns (google.protobuf.Empty);
+	rpc TtyWinResize(TtyWinResizeRequest) returns (google.protobuf.Empty);
+
+	// misc (TODO: some rpcs can be replaced by hyperstart-exec)
+	rpc StartSandbox(StartSandboxRequest) returns (google.protobuf.Empty);
+	rpc DestroySandbox(DestroySandboxRequest) returns (google.protobuf.Empty);
+	rpc UpdateInterface(UpdateInterfaceRequest) returns (google.protobuf.Empty);
+	rpc AddRoute(AddRouteRequest) returns (google.protobuf.Empty);
+	rpc OnlineCPUMem(OnlineCPUMemRequest) returns (google.protobuf.Empty);
+}
+
+message CreateContainerRequest {
+	string container_id = 1;
+	Spec OCI = 2;
+}
+
+message StartContainerRequest {
+	string container_id = 1;
+}
+
+message ExecProcessRequest {
+	string container_id = 1;
```

Comment:
add StringUser here

Reply:
@laijs This would be redundant with the `Process.User` field, wouldn't it ?

Reply:
quote from comments by @laijs above

> StringUser is needed since the oci spec don't allow string user ids

correct this if anything wrong

Reply:
Sure, to make container boot quickly, avoiding mounting the root block device on the host is very important.

Reply:
@gnawuc @laijs the OCI spec passes a uint32 for both UID and GID, but that should be enough. Why would a uint32 not be good enough for a UID/GID and why would we absolutely need a string?

Reply:
@sameo what @laijs defined here is string user like `gnawux` or `sameo`, user may specify a username instead of a number id, and we couldn't get the id number from the `/etc/passwd` without mounting it if a block device is passed for rootfs.

Reply:
@gnawux @laijs About this, I am not sure if we still need this. First, the OCI spec defines a `User` structure with the field `Username` in it https://github.com/kata-containers/agent/blob/master/protocols/grpc/oci.pb.go#L267. This field is a string as expected.
This would cover partially what we are doing with `StringUser`.

But I also wanted to talk about the root cause of the problem here. I would like to understand why we cannot mount `/etc/passwd` inside the container ? Even in case of rootfs being a block device, we can use our `9pfs` shared directory to bind mount `/etc/passwd` from the host to the guest, and then libcontainer would mount `/storage/abc/etc/password` into `/etc/passwd` inside our container.
Does this make sense or am I missing something here ?

Reply:
@sboeuf That usename is used for windows only: https://github.com/opencontainers/runtime-spec/commit/f9e48e0024372f1abcc6271a88328c0e72ffbaf6

It is a little weird if we reuse this variable. @gnawux do you accept it ?

Reply:
We can ways mount the block device inside the container to access to the `/etc/passwd`.

But we can't do mount the block device in the host (for accessing to the `/etc/passwd` and filling the numerical user id in the spec).

1) the host might not have the ability to mount it. for example, if the block device is from ceph, and the host doesn't have krbd.ko. And even the host has it, userspace ceph lib + qume are alway the best choice.

2) Security, if the block device was once mounted inside the vm, the host should not mount it. the code in the vm might hack into the guest kernel, and modify the metadata of the filesystem of the block device. The host kernel might be also broken into when mounting the block device.

3) performance: to speed up the starting of the container, hyper avoid mounting the block device on the host.

Reply:
@laijs not sure I follow everything here because I thought that `/etc/passwd` was passed through `9pfs` through the shared directory. And in case `/etc/passwd` comes from the block device passed through the VM, it is mounted somewhere, meaning that you can specify libcontainer that you expect `/etc/passwd` to be mounted, right ?

Reply:
@laijs just to be clear, I am not saying that I want to mount the block device on the host, I agree it would take some time. We don't do that either.
But the block device needs to be mounted inside the guest by the agent and then we can provide libcontainer with the right mounts so that it will mount `/etc/passwd` in the right place. This way, it would be able to know what `1000:1000` means for the user/group.

Reply:
You are right.
Runtime --> agent --> libcontainer.
Runtime has to pass the string user name to agent if string name is specified by upper layer tools. Agent parses the string name from `/etc/passwd` and pass the digit user id to the libcontainer.
"Runtime --> agent" is the API, so we need the string user name in the API.

docker-runc does't allow the string user name, so the docker/moby does the parsing of the username. kata runtime cli will not allow the string user name either. But kata runtime can also be a lib for hyperd/frakti, the string user name will be passed to the agent in this case if requested.